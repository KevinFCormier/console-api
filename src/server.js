/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import https from 'https';
import fs from 'fs';
import schema from './schema/index';
import config from '../config';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const GRAPHQL_PORT = process.env.PORT || config.get('httpPort') || 4000;
const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(req => ({ formatError, schema, context: req })));

const privateKey = fs.readFileSync(process.env.serverKey || './sslcert/hcmuiapi.key', 'utf8');
const certificate = fs.readFileSync(process.env.serverCert || './sslcert/hcmuiapi.crt', 'utf8');
const credentials = { key: privateKey, cert: certificate };
const server = https.createServer(credentials, graphQLServer);

server.listen(GRAPHQL_PORT, () => {
  logger.info(`[pid ${process.pid}] [env ${process.env.NODE_ENV}] started.`);
  logger.info(`HCM UI API is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphql`);
  if (process.env.NODE_ENV !== 'production') {
    logger.info(`GraphiQL is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphiql`);
  }
});
