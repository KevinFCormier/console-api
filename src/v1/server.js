/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import schema from './schema/index';
import config from '../../config';
import HCMConnector from './datasource/lib/hcm-connector';
import MongoConnector from './datasource/lib/mongo-connector';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};


const getProductionOptions = async (req) => {
  const mongoConnector = new MongoConnector();
  const hcmConnector = new HCMConnector({
    hcmUrl: config.get('hcmUrl') || 'http://localhost:8080',
    pollInterval: config.get('hcmPollInterval') || 200,
    pollTimeout: config.get('hcmPollTimeout') || 10000,
  });

  await mongoConnector.connect(config.get('mongodbUrl') || 'mongodb://localhost:27017/weave');
  return {
    formatError,
    schema,
    context: {
      req,
      hcmConnector,
      mongoConnector,
    },
  };
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(getProductionOptions));

export default graphQLServer;
