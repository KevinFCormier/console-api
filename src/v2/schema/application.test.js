/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import supertest from 'supertest';
import server, { GRAPHQL_PATH } from '../index';

describe('Application Resolver', () => {
  test('Correctly Resolves Applications Query', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          applications {
            selector
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Single Applications Query', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          applications(name:"app02",namespace:"default") {
            selector
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });


  test('Correctly Resolves Deployables Query - MatchNames', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          deployables(selector:{
            matchNames:[{
              name: "deployable01"
            }]
          }) {
            deployer {
              chartName
              namespace
              repository
              version
            }
            dependencies {
              kind
              name
            }
            metadata {
              name
            }
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Placement Policies Query', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          placementPolicies(selector:{
             matchNames:[{
              name:"placement02"
            }]
          }) {
            clusterSelector
            metadata {
              annotations
              name
              namespace
            }
            replicas
            resourceSelector
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Create Application Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          createApplication(resources:[{
            apiVersion: "mcm.ibm.com/v1alpha1",
            kind: "Application",
            metadata: {
              name: "testapp",
              labels: {
                deployable: "deployable01",
                hcmapp: "testapp",
              },
            },
            spec: {
              selector: {
                matchLabels: {
                  hcmapp: "testapp",
                },
              },
              componentKinds: [
                {
                  group: "mcm.ibm.com/v1alpha1",
                  kind: "PlacementPolicy",
                },
                {
                  group: "mcm.ibm.com/v1alpha1",
                  kind: "Deployable",
                },
              ],
            },
          }])
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Delete Application Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          deleteApplication(name:"testapp",namespace:"default")
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });
});

