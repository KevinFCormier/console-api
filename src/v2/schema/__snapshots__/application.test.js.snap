// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application Resolver Correctly Resolves Applications Query 1`] = `
Object {
  "data": Object {
    "applications": Array [
      Object {
        "selector": Object {
          "matchLabels": Object {
            "hcmapp": "app01",
          },
        },
      },
      Object {
        "selector": Object {
          "matchLabels": Object {
            "hcmapp": "app02",
          },
        },
      },
      Object {
        "selector": Object {
          "matchLabels": Object {
            "hcmapp": "app01",
          },
        },
      },
      Object {
        "selector": Object {
          "matchLabels": Object {
            "hcmapp": "app02",
          },
        },
      },
    ],
  },
}
`;

exports[`Application Resolver Correctly Resolves Create Application Mutation 1`] = `
Object {
  "data": Object {
    "createApplication": Object {
      "errors": Array [],
      "result": Array [
        Object {
          "apiVersion": "mcm.ibm.com/v1alpha1",
          "kind": "Application",
          "metadata": Object {
            "annotations": Object {
              "mcm.ibm.com/user-group": "c3lzdGVtOmF1dGhlbnRpY2F0ZWQ=",
              "mcm.ibm.com/user-identity": "aHR0cHM6Ly9teWNsdXN0ZXIuaWNwOjk0NDMvb2lkYy9lbmRwb2ludC9PUCNhZG1pbg==",
            },
            "creationTimestamp": "2018-09-07T14:19:02Z",
            "labels": Object {
              "deployable": "deployable01",
              "hcmapp": "testapp",
            },
            "name": "testapp",
            "namespace": "default",
            "resourceVersion": "59905",
            "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/default/applications/testapp",
            "uid": "f849165a-b2a8-11e8-bd43-b69970856045",
          },
          "spec": Object {
            "componentKinds": Array [
              Object {
                "group": "mcm.ibm.com/v1alpha1",
                "kind": "PlacementPolicy",
              },
              Object {
                "group": "mcm.ibm.com/v1alpha1",
                "kind": "Deployable",
              },
            ],
            "descriptor": Object {},
            "selector": Object {
              "matchLabels": Object {
                "hcmapp": "testapp",
              },
            },
          },
          "status": Object {
            "Deployable": Object {
              "metadata": Object {
                "creationTimestamp": null,
              },
              "spec": Object {
                "deployer": Object {
                  "helm": Object {},
                },
              },
              "status": Object {},
            },
          },
        },
      ],
    },
  },
}
`;

exports[`Application Resolver Correctly Resolves Delete Application Mutation 1`] = `
Object {
  "data": Object {
    "deleteApplication": "testapp",
  },
}
`;

exports[`Application Resolver Correctly Resolves Deployables Query - MatchLabels 1`] = `
Object {
  "data": Object {
    "deployables": Array [
      Object {
        "dependencies": null,
        "deployer": Object {
          "chartName": "nginx-lego",
          "namespace": "default",
          "repository": "google",
          "version": "0.3.1",
        },
        "metadata": Object {
          "name": "deployable02-a",
        },
      },
      Object {
        "dependencies": null,
        "deployer": Object {
          "chartName": "nginx-lego",
          "namespace": "default",
          "repository": "google",
          "version": "0.3.1",
        },
        "metadata": Object {
          "name": "deployable02-b",
        },
      },
      Object {
        "dependencies": null,
        "deployer": Object {
          "chartName": "nginx-lego",
          "namespace": "default",
          "repository": "google",
          "version": "0.3.1",
        },
        "metadata": Object {
          "name": "deployable02-a",
        },
      },
      Object {
        "dependencies": null,
        "deployer": Object {
          "chartName": "nginx-lego",
          "namespace": "default",
          "repository": "google",
          "version": "0.3.1",
        },
        "metadata": Object {
          "name": "deployable02-b",
        },
      },
    ],
  },
}
`;

exports[`Application Resolver Correctly Resolves Placement Policies Query 1`] = `
Object {
  "data": Object {
    "placementPolicies": Array [
      Object {
        "clusterSelector": Object {
          "matchLabels": Object {
            "name": "mycluster.icp",
          },
        },
        "metadata": Object {
          "annotations": Object {
            "mcm.ibm.com/user-group": "c3lzdGVtOmF1dGhlbnRpY2F0ZWQ=",
            "mcm.ibm.com/user-identity": "aHR0cHM6Ly9teWNsdXN0ZXIuaWNwOjk0NDMvb2lkYy9lbmRwb2ludC9PUCNhZG1pbg==",
          },
          "name": "placement02",
          "namespace": "default",
        },
        "replicas": 1,
        "resourceSelector": Object {},
      },
      Object {
        "clusterSelector": Object {
          "matchLabels": Object {
            "name": "mycluster.icp",
          },
        },
        "metadata": Object {
          "annotations": Object {
            "mcm.ibm.com/user-group": "c3lzdGVtOmF1dGhlbnRpY2F0ZWQ=",
            "mcm.ibm.com/user-identity": "aHR0cHM6Ly9teWNsdXN0ZXIuaWNwOjk0NDMvb2lkYy9lbmRwb2ludC9PUCNhZG1pbg==",
          },
          "name": "placement02",
          "namespace": "default",
        },
        "replicas": 1,
        "resourceSelector": Object {},
      },
    ],
  },
}
`;

exports[`Application Resolver Correctly Resolves Single Applications Query 1`] = `
Object {
  "data": Object {
    "applications": Array [
      Object {
        "selector": Object {
          "matchLabels": Object {
            "hcmapp": "app02",
          },
        },
      },
    ],
  },
}
`;
